### ISSUE TITLE
Simulator state shouldn't expose external details

### ISSUE THREAD

**Issue by Strilanc** [quantumlib/Cirq#1658]
```
import cirq
d = cirq.asymmetric_depolarize(0.1, 0, 0)
s = cirq.DensityMatrixSimulator()
a, b = cirq.LineQubit.range(2)
out = s.simulate(cirq.Circuit.from_ops(d(a), d(b)))
assert out.final_simulator_state.density_matrix.shape == (4, 4)
```

Fails saying the shape is `(2, 2, 2, 2)`. If you use an empty circuit (and specify the qubits) you instead get shape `(4, 4)`.

**Comment by dabacon** [comment]
Good catch. 

I think numpy is smart enough when you reshape tensors, but I assume we should always do (n, n) for density matrix or (n,) for state.  Should we consider adding versions of these that return the tensor form of these?

**Comment by Strilanc** [comment]
I think we should always do `(n, n)` for now.

**Comment by dabacon** [comment]
Ah note that you are doing `out.final_simulator_state`.  That's the internal "simulator state".  The right way to get this is `out.final_density_matrix`.

**Comment by dabacon** [comment]
So yeah the issue here is that the public facing stuff should be in trial results and step results, but not in the "simulator state".  That way you can't do it the "wrong way" without considerable effort and using "_" classes.

**Comment by dabacon** [comment]
Fixed in #1678
